File: internal/llm/openai.go
package llm

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"
)

type OpenAIProvider struct {
	name     string
	apiKey   string
	baseURL  string
	model    string
	client   *http.Client
}

func NewOpenAIProvider(apiKey, baseURL, model string) *OpenAIProvider {
	if baseURL == "" {
		baseURL = "https://api.openai.com/v1"
	}

	return &OpenAIProvider{
		name:    "openai",
		apiKey:  apiKey,
		baseURL: baseURL,
		model:   model,
		client: &http.Client{
			Timeout: 60 * time.Second,
			Transport: &http.Transport{
				MaxIdleConns:    20,
				IdleConnTimeout: 30 * time.Second,
			},
		},
	}
}

func (p *OpenAIProvider) Name() string { return p.name }

func (p *OpenAIProvider) IsAvailable() bool {
	return p.apiKey != ""
}

type openAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type openAIRequest struct {
	Model       string          `json:"model"`
	Messages    []openAIMessage `json:"messages"`
	MaxTokens   int             `json:"max_tokens,omitempty"`
	Temperature float64         `json:"temperature,omitempty"`
	Stream      bool            `json:"stream,omitempty"`
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage"`
	Model string `json:"model"`
	Error *struct {
		Message string `json:"message"`
		Code    string `json:"code"`
	} `json:"error,omitempty"`
}

func (p *OpenAIProvider) Chat(ctx context.Context, req *Request) (*Response, error) {
	messages := make([]openAIMessage, len(req.Messages))
	for i, m := range req.Messages {
		messages[i] = openAIMessage{Role: m.Role, Content: m.Content}
	}

	body := openAIRequest{
		Model:       p.model,
		Messages:    messages,
		MaxTokens:   req.MaxTokens,
		Temperature: req.Temperature,
	}

	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", p.baseURL+"/chat/completions", bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("http request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode != 200 {
		slog.Warn("openai error", "status", resp.StatusCode, "body", string(respBody[:min(len(respBody), 500)]))
		// Try to parse error
		var errResp openAIResponse
		if json.Unmarshal(respBody, &errResp) == nil && errResp.Error != nil {
			return nil, fmt.Errorf("openai: %s (%s)", errResp.Error.Message, errResp.Error.Code)
		}
		return nil, fmt.Errorf("openai: HTTP %d", resp.StatusCode)
	}

	var result openAIResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}

	if len(result.Choices) == 0 {
		return nil, fmt.Errorf("openai: no choices returned")
	}

	return &Response{
		Content:      result.Choices[0].Message.Content,
		Model:        result.Model,
		InputTokens:  result.Usage.PromptTokens,
		OutputTokens: result.Usage.CompletionTokens,
	}, nil
}

// openAIStreamChunk represents a single SSE delta from a streaming response.
type openAIStreamChunk struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Model   string `json:"model"`
	Choices []struct {
		Index int `json:"index"`
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
		FinishReason *string `json:"finish_reason"`
	} `json:"choices"`
	Usage *struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage,omitempty"`
}

func (p *OpenAIProvider) ChatStream(ctx context.Context, req *Request, onChunk func(string)) (*Response, error) {
	messages := make([]openAIMessage, len(req.Messages))
	for i, m := range req.Messages {
		messages[i] = openAIMessage{Role: m.Role, Content: m.Content}
	}

	body := openAIRequest{
		Model:       p.model,
		Messages:    messages,
		MaxTokens:   req.MaxTokens,
		Temperature: req.Temperature,
		Stream:      true,
	}

	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", p.baseURL+"/chat/completions", bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", "Bearer "+p.apiKey)
	httpReq.Header.Set("Accept", "text/event-stream")

	resp, err := p.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("http request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		respBody, _ := io.ReadAll(resp.Body)
		slog.Warn("openai stream error", "status", resp.StatusCode, "body", string(respBody[:min(len(respBody), 500)]))
		return nil, fmt.Errorf("openai: HTTP %d", resp.StatusCode)
	}

	result := &Response{
		Model: p.model,
	}

	var fullContent string
	br := bufio.NewReader(resp.Body)
	for {
		line, err := br.ReadString('\n')
		if err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return nil, fmt.Errorf("read stream: %w", err)
		}

		line = strings.TrimSpace(line)

		// Skip empty lines and keep-alive comments
		if line == "" || strings.HasPrefix(line, ":") {
			continue
		}

		// Expect "data: ..."
		if !strings.HasPrefix(line, "data: ") {
			continue
		}

		payload := strings.TrimPrefix(line, "data: ")

		// Stream end signal
		if payload == "[DONE]" {
			break
		}

		var chunk openAIStreamChunk
		if err := json.Unmarshal([]byte(payload), &chunk); err != nil {
			slog.Debug("openai stream parse error", "error", err, "line", line)
			continue
		}

		if len(chunk.Choices) > 0 {
			content := chunk.Choices[0].Delta.Content
			if content != "" {
				fullContent += content
				onChunk(content)
			}

			// Capture usage from the last chunk if present
			if chunk.Usage != nil {
				result.InputTokens = chunk.Usage.PromptTokens
				result.OutputTokens = chunk.Usage.CompletionTokens
			}

			// Also capture model from any chunk
			if chunk.Model != "" {
				result.Model = chunk.Model
			}
		}
	}

	result.Content = fullContent
	return result, nil
}

File: python-engine/app/providers/openai.py
# OpenAI Provider — 使用 openai SDK
from __future__ import annotations

import json
import logging
from typing import AsyncIterator

from openai import AsyncOpenAI

from app.gateway.provider import (
    ChatMessage,
    ChatResponse,
    EmbeddingResponse,
    LLMProvider,
    ToolCall,
)

logger = logging.getLogger(__name__)


class OpenAIProvider(LLMProvider):
    name = "openai"

    def __init__(self, api_key: str, base_url: str = ""):
        kwargs: dict = {"api_key": api_key}
        if base_url:
            kwargs["base_url"] = base_url
        self._client = AsyncOpenAI(**kwargs)

    async def chat_stream(
        self,
        messages: list[ChatMessage],
        model: str,
        *,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        tools: list[dict] | None = None,
    ) -> AsyncIterator[ChatResponse]:
        kwargs = self._build_kwargs(messages, model, max_tokens, temperature, tools)
        kwargs["stream"] = True

        response = await self._client.chat.completions.create(**kwargs)

        tool_calls: list[dict] = []
        input_tokens = 0
        output_tokens = 0
        reasoning_content = ""

        async for chunk in response:
            if not chunk.choices:
                # 最后一个 chunk 可能只有 usage
                if chunk.usage:
                    input_tokens = chunk.usage.prompt_tokens
                    output_tokens = chunk.usage.completion_tokens
                continue

            delta = chunk.choices[0].delta

            # DeepSeek thinking mode：获取 reasoning_content（API 发送的是增量文本）
            if hasattr(delta, 'reasoning_content') and delta.reasoning_content:
                reasoning_content += delta.reasoning_content  # 累积完整思考文本
                yield ChatResponse(reasoning_content=delta.reasoning_content)  # 传递增量

            if delta.content:
                yield ChatResponse(content=delta.content)

            if delta.tool_calls:
                for tc in delta.tool_calls:
                    while len(tool_calls) <= tc.index:
                        tool_calls.append({"id": "", "name": "", "arguments": ""})
                    if tc.id:
                        tool_calls[tc.index]["id"] = tc.id
                    if tc.function:
                        if tc.function.name:
                            tool_calls[tc.index]["name"] = tc.function.name
                        if tc.function.arguments:
                            tool_calls[tc.index]["arguments"] += tc.function.arguments

            finish = chunk.choices[0].finish_reason
            if finish:
                # 捕获 Token 计数（可能和 finish_reason 在同一 chunk）
                if chunk.usage:
                    input_tokens = chunk.usage.prompt_tokens
                    output_tokens = chunk.usage.completion_tokens
                parsed = [
                    ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"])
                    for tc in tool_calls if tc["id"]
                ]
                yield ChatResponse(
                    tool_calls=parsed,
                    finish_reason="tool_calls" if parsed else (finish or "stop"),
                    input_tokens=input_tokens,
                    output_tokens=output_tokens,
                    reasoning_content=reasoning_content,
                )

        # 发出 token 统计（usage chunk 可能没有 finish_reason）
        if input_tokens > 0 or output_tokens > 0:
            yield ChatResponse(
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                finish_reason="stop",
                reasoning_content=reasoning_content,
            )

    async def chat(
        self,
        messages: list[ChatMessage],
        model: str,
        *,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        tools: list[dict] | None = None,
    ) -> ChatResponse:
        kwargs = self._build_kwargs(messages, model, max_tokens, temperature, tools)
        resp = await self._client.chat.completions.create(**kwargs)

        choice = resp.choices[0]
        content = choice.message.content or ""
        tool_calls = []
        if choice.message.tool_calls:
            for tc in choice.message.tool_calls:
                tool_calls.append(
                    ToolCall(id=tc.id, name=tc.function.name, arguments=tc.function.arguments)
                )

        usage = resp.usage
        return ChatResponse(
            content=content,
            tool_calls=tool_calls,
            finish_reason="tool_calls" if tool_calls else (choice.finish_reason or "stop"),
            input_tokens=usage.prompt_tokens if usage else 0,
            output_tokens=usage.completion_tokens if usage else 0,
        )

    async def embed(self, text: str, model: str) -> EmbeddingResponse:
        resp = await self._client.embeddings.create(model=model, input=text)
        usage = resp.usage
        return EmbeddingResponse(
            embedding=resp.data[0].embedding,
            tokens=usage.total_tokens if usage else 0,
        )

    async def close(self) -> None:
        await self._client.close()

    # ── helpers ──

    @staticmethod
    def _build_kwargs(
        messages: list[ChatMessage],
        model: str,
        max_tokens: int,
        temperature: float,
        tools: list[dict] | None,
    ) -> dict:
        kwargs: dict = {
            "model": model,
            "messages": [m.to_dict() if hasattr(m, 'to_dict') else m for m in messages],
            "max_tokens": max_tokens,
            "temperature": temperature,
        }
        # ── 调试：检查消息序列中 tool 消息的配对 ──
        msgs = kwargs["messages"]
        for i, m in enumerate(msgs):
            if m.get("role") == "tool":
                prev = msgs[i-1] if i > 0 else None
                if prev:
                    has_tc = bool(prev.get("tool_calls"))
                    logger.info("Tool msg #%d: prev role=%s has_tool_calls=%s tool_call_id=%s",
                                i, prev.get("role"), has_tc, m.get("tool_call_id", ""))
        logger.info("LLM call: model=%s msgs=%d roles=%s", model, len(msgs), [m.get("role") for m in msgs])
        if tools:
            # 统一转为 OpenAI function 格式
            converted = []
            for t in tools:
                if "type" in t and t["type"] == "function":
                    converted.append(t)
                else:
                    converted.append({"type": "function", "function": t})
            kwargs["tools"] = converted
            kwargs["tool_choice"] = "auto"
        return kwargs
